zrml/prediction-markets/src/lib.rs

```diff
@@ -599,7 +610,22 @@ mod pallet {
             // TODO(#787): Handle Rikiddo benchmarks!
             let sender = ensure_signed(origin)?;
 
+            let valid_base_asset = match base_asset {
+                Asset::Ztg => Ok::<bool, Error<T>>(true),
+                #[cfg(feature = "parachain")]
+                Asset::ForeignAsset(fa) => {
+                    if let Some(metadata) = T::AssetRegistry::metadata(&Asset::ForeignAsset(fa)) {
+                        Ok::<bool, Error<T>>(metadata.additional.allow_as_base_asset)
+                    } else {
+                        Err::<bool, Error<T>>(Error::<T>::UnregisteredForeignAsset)
+                    }
+                }
+                _ => Ok(false),
+            }?;
```

Comment:
That looks kind of strange. If `valid_base_asset` contains the Error `UnregisteredForeignAsset`, what does happen when this reaches the below ensure check? I think this is quite confusing. 

```suggestion
            let mut valid_base_asset = false;
            match base_asset {
                Asset::Ztg => valid_base_asset = true,
                #[cfg(feature = "parachain")]
                Asset::ForeignAsset(fa) => {
                    let metadata = T::AssetRegistry::metadata(&Asset::ForeignAsset(fa)).ok_or(Error::<T>::UnregisteredForeignAsset);
                    valid_base_asset = metadata..additional.allow_as_base_asset;
                }
                _ => valid_base_asset = false,
            }?;
```

Reply:
I am sorry. I did not see [that.](https://github.com/zeitgeistpm/zeitgeist/pull/894#discussion_r1048946784) But the style with error seems still a little bit strange.

Reply:
? on line 622 will make function return error and it never reaches ensure on line 626. And test is updated to cover this case.

Reply:
earlier it was like what @Chralt98 suggested but after @maltekliemann 's suggestion I changed it. I find both options fine.

Reply:
It's just that I don't think that the following syntax is well understood at the first glance.

```rust
Ok::<bool, Error<T>>(true)
```

So I would like to stick with the "readable" version with `mut valid_base_asset`, but it's not that important. It just was so noticeable.

Reply:
Chris is absolutely right, that's not pretty. But if you just add a type hint to `valid_base_asset`, we can get rid of the other type hints.

```rust
            let valid_base_asset: bool = match base_asset {
                Asset::Ztg => Ok(true),
                #[cfg(feature = "parachain")]
                Asset::ForeignAsset(fa) => {
                    if let Some(metadata) = T::AssetRegistry::metadata(&Asset::ForeignAsset(fa)) {
                        Ok(metadata.additional.allow_as_base_asset)
                    } else {
                        Err(Error::<T>::UnregisteredForeignAsset)
                    }
                }
                _ => Ok(false),
            }?;
```

Reply:
You can also skip the entire result and just return an error where appropriate:

```rust
            let valid_base_asset = match base_asset {
                Asset::Ztg => true,
                #[cfg(feature = "parachain")]
                Asset::ForeignAsset(fa) => {
                    if let Some(metadata) = T::AssetRegistry::metadata(&Asset::ForeignAsset(fa)) {
                        metadata.additional.allow_as_base_asset
                    } else {
                        return Err(Error::<T>::UnregisteredForeignAsset.into());
                    }
                }
                _ => false,
            };
```